// ts-colors/lib/HSVA.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fromHSLA = exports.hsv = exports.hsva = void 0;
var Hue_1 = require("./Hue");
var UnitInterval_1 = require("./UnitInterval");
/**
 * @since 0.1.5
 * @category constructors
 */
var hsva = function (h, s, v, a) { return ({
    h: Hue_1.clipHue(h),
    s: UnitInterval_1.unitInterval(s),
    v: UnitInterval_1.unitInterval(v),
    a: UnitInterval_1.unitInterval(a)
}); };
exports.hsva = hsva;
/**
 * @since 0.1.5
 * @category constructors
 */
var hsv = function (h, s, v) { return exports.hsva(h, s, v, 1.0); };
exports.hsv = hsv;
/**
 * @since 0.1.5
 * @category constructors
 */
var fromHSLA = function (_a) {
    var h = _a.h, s = _a.s, l = _a.l, a = _a.a;
    var tmp = s * (l < 0.5 ? l : 1.0 - l);
    var hue = Hue_1.clipHue(h);
    var saturation = (2.0 * tmp) / (l + tmp);
    var v = l + tmp;
    if (l === 0) {
        return exports.hsva(hue, (2.0 * s) / (1.0 + s), 0.0, a);
    }
    if (s === 0 && l === 1) {
        return exports.hsva(hue, 0.0, 1.0, a);
    }
    return exports.hsva(hue, saturation, v, a);
};
exports.fromHSLA = fromHSLA;
