// ts-colors/lib/Lab.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mix = exports.evolve = exports.fromLCh = exports.fromHSLA = exports.lab = void 0;
/**
 * A `Color` represented L, a and b coordinates in the Lab color space.
 *
 * Note: See documentation for [`xyz`](./XYZ.ts.html). The same restrictions
 * apply here.
 *
 * @since 0.1.5
 */
var function_1 = require("fp-ts/function");
var struct = __importStar(require("fp-ts/struct"));
var XYZ = __importStar(require("./XYZ"));
var math_1 = require("./math");
/**
 * @category constructors
 * @since 0.1.5
 */
var lab = function (l, a, b) { return ({
    l: l,
    a: a,
    b: b
}); };
exports.lab = lab;
/**
 * @since 0.1.5
 * @category constructors
 */
var fromHSLA = function (c) {
    var cut = Math.pow(6.0 / 29.0, 3.0);
    var f = function (t) {
        return t > cut
            ? Math.pow(t, 1.0 / 3.0)
            : (1.0 / 3.0) * Math.pow(29.0 / 6.0, 2.0) * t + 4.0 / 29.0;
    };
    var rec = XYZ.fromHSLA(c);
    var fy = f(rec.y / XYZ.D65.yn);
    var l = 116.0 * fy - 16.0;
    var a = 500.0 * (f(rec.x / XYZ.D65.xn) - fy);
    var b = 200.0 * (fy - f(rec.z / XYZ.D65.zn));
    return exports.lab(l, a, b);
};
exports.fromHSLA = fromHSLA;
/**
 * @category constructors
 * @since 0.1.5
 */
var fromLCh = function (_a) {
    var l = _a.l, c = _a.c, h = _a.h;
    var a = c * Math.cos(h * math_1.deg2rad);
    var b = c * Math.sin(h * math_1.deg2rad);
    return exports.lab(l, a, b);
};
exports.fromLCh = fromLCh;
/**
 * @since 0.1.5
 */
var evolve = function (t) { return function (c) {
    return function_1.pipe(c, struct.evolve(t), function (_a) {
        var l = _a.l, a = _a.a, b = _a.b;
        return exports.lab(l, a, b);
    });
}; };
exports.evolve = evolve;
/**
 * @since 0.1.5
 */
var mix = function (ratio) {
    return function (a) {
        return function (b) {
            var i = math_1.interpolate(ratio);
            return function_1.pipe(b, exports.evolve({
                l: i(a.l),
                a: i(a.a),
                b: i(a.b)
            }));
        };
    };
};
exports.mix = mix;
