// ts-colors/lib/RGBA.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Show = exports.Eq = exports.toCSS = exports.mix = exports.evolve = exports.luminance = exports.brightness = exports.chroma = exports.minChroma = exports.maxChroma = exports.fromXYZ = exports.fromHSLA = exports.fromNormalized = exports.normalizedFromHSLA = exports.normalize = exports.normalized = exports.rgb = exports.rgba = exports.normalizedChannel = exports.denormalizeChannel = exports.channel = void 0;
/**
 * @since 0.1.5
 */
var Ord = __importStar(require("fp-ts/Ord"));
var Equals = __importStar(require("fp-ts/Eq"));
var Sh = __importStar(require("fp-ts/Show"));
var number = __importStar(require("fp-ts/number"));
var struct = __importStar(require("fp-ts/struct"));
var UnitInterval_1 = require("./UnitInterval");
var math_1 = require("./math");
var function_1 = require("fp-ts/function");
var clampChannel = Ord.clamp(number.Ord)(0, 255);
/**
 * @since 0.1.5
 * @internal
 */
var channel = function (n) {
    return function_1.pipe(Math.round(n), clampChannel);
};
exports.channel = channel;
/**
 * @since 0.1.5
 * @internal
 */
var denormalizeChannel = function (n) {
    return function_1.pipe(Math.round(n * 255.0), exports.channel);
};
exports.denormalizeChannel = denormalizeChannel;
/**
 * @since 0.1.5
 * @internal
 */
exports.normalizedChannel = UnitInterval_1.unitInterval;
/**
 * @since 0.1.5
 * @category constructors
 */
var rgba = function (r, g, b, a) { return ({
    r: exports.channel(r),
    g: exports.channel(g),
    b: exports.channel(b),
    a: UnitInterval_1.unitInterval(a)
}); };
exports.rgba = rgba;
/**
 * @since 0.1.5
 * @category constructors
 */
var rgb = function (r, g, b) { return exports.rgba(r, g, b, 1.0); };
exports.rgb = rgb;
/**
 * @since 0.1.5
 * @category constructors
 * @internal
 */
var normalized = function (r, g, b, a) { return ({
    r: exports.normalizedChannel(r),
    g: exports.normalizedChannel(g),
    b: exports.normalizedChannel(b),
    a: UnitInterval_1.unitInterval(a)
}); };
exports.normalized = normalized;
/**
 * @since 0.1.5
 * @category constructors
 * @internal
 */
var normalize = function (c) {
    return exports.normalized(c.r / 255, c.g / 255, c.b / 255, c.a);
};
exports.normalize = normalize;
/**
 * @since 0.1.5
 * @category constructors
 * @internal
 */
var normalizedFromHSLA = function (_a) {
    var h = _a.h, s = _a.s, l = _a.l, a = _a.a;
    var ch = h / 60.0;
    var chr = (1.0 - Math.abs(2.0 * l - 1.0)) * s;
    var m = l - chr / 2.0;
    var x = chr * (1.0 - Math.abs((ch % 2.0) - 1.0));
    var channels = function () {
        if (ch < 1.0) {
            return { r: chr, g: x, b: 0.0 };
        }
        if (1.0 <= ch && ch < 2.0) {
            return { r: x, g: chr, b: 0.0 };
        }
        if (2.0 <= ch && ch < 3.0) {
            return { r: 0.0, g: chr, b: x };
        }
        if (3.0 <= ch && ch < 4.0) {
            return { r: 0.0, g: x, b: chr };
        }
        if (4.0 <= ch && ch < 5.0) {
            return { r: x, g: 0.0, b: chr };
        }
        return { r: chr, g: 0.0, b: x };
    };
    var rgb = channels();
    return exports.normalized(rgb.r + m, rgb.g + m, rgb.b + m, a);
};
exports.normalizedFromHSLA = normalizedFromHSLA;
/**
 * @since 0.1.5
 * @category constructors
 * @internal
 */
var fromNormalized = function (c) {
    return exports.rgba(exports.denormalizeChannel(c.r), exports.denormalizeChannel(c.g), exports.denormalizeChannel(c.b), c.a);
};
exports.fromNormalized = fromNormalized;
/**
 * @since 0.1.5
 * @category constructors
 */
var fromHSLA = function (c) {
    return function_1.pipe(exports.normalizedFromHSLA(c), exports.fromNormalized);
};
exports.fromHSLA = fromHSLA;
/**
 * @category constructors
 * @since 0.1.5
 */
var fromXYZ = function (_a) {
    var x = _a.x, y = _a.y, z = _a.z;
    var f = function (c) {
        return c <= 0.0031308 ? 12.92 * c : 1.055 * Math.pow(c, 1.0 / 2.4) - 0.055;
    };
    var r = f(3.2406 * x - 1.5372 * y - 0.4986 * z);
    var g = f(-0.9689 * x + 1.8758 * y + 0.0415 * z);
    var b = f(0.0557 * x - 0.204 * y + 1.057 * z);
    return exports.rgb(r * 255, g * 255, b * 255);
};
exports.fromXYZ = fromXYZ;
/**
 * @since 0.1.5
 */
var maxChroma = function (_a) {
    var r = _a.r, g = _a.g, b = _a.b;
    return Math.max(Math.max(r, g), b);
};
exports.maxChroma = maxChroma;
/**
 * @since 0.1.5
 */
var minChroma = function (_a) {
    var r = _a.r, g = _a.g, b = _a.b;
    return Math.min(Math.min(r, g), b);
};
exports.minChroma = minChroma;
/**
 * @since 0.1.5
 */
var chroma = function (c) { return exports.maxChroma(c) - exports.minChroma(c); };
exports.chroma = chroma;
/**
 * The percieved brightness of the color (A number between 0.0 and 1.0).
 * See: [https://www.w3.org/TR/AERT#color-contrast](https://www.w3.org/TR/AERT#color-contrast)
 *
 * @since 0.1.5
 */
var brightness = function (c) {
    return (299.0 * c.r + 587.0 * c.g + 114.0 * c.b) / 1000.0;
};
exports.brightness = brightness;
/**
 * The relative brightness of a color (normalized to 0.0 for darkest black
 * and 1.0 for lightest white), according to the WCAG definition.
 *
 * See: [https://www.w3.org/TR/2008/REC-WCAG20-20081211/#relativeluminancedef](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#relativeluminancedef)
 *
 * @since 0.1.5
 */
var luminance = function (rgba) {
    // const rgba = normalize(c)
    var f = function (c) {
        if (c <= 0.03928) {
            return c / 12.92;
        }
        return Math.pow((c + 0.055) / 1.055, 2.4);
    };
    var r = f(rgba.r);
    var g = f(rgba.g);
    var b = f(rgba.b);
    return 0.2126 * r + 0.7152 * g + 0.0722 * b;
};
exports.luminance = luminance;
/**
 * @since 0.1.5
 */
var evolve = function (t) { return function (c) {
    return function_1.pipe(c, struct.evolve(t), function (_a) {
        var r = _a.r, g = _a.g, b = _a.b, a = _a.a;
        return exports.rgba(r, g, b, a);
    });
}; };
exports.evolve = evolve;
/**
 * @since 0.1.5
 */
var mix = function (ratio) {
    return function (a) {
        return function (b) {
            var i = math_1.interpolate(ratio);
            return function_1.pipe(b, exports.evolve({
                r: i(a.r),
                g: i(a.g),
                b: i(a.b),
                a: i(a.a)
            }));
        };
    };
};
exports.mix = mix;
/**
 * A CSS representation of the color in the form `rgb(..)` or `rgba(...)`
 *
 * @since 0.1.5
 * @category destructors
 */
var toCSS = function (c) {
    return c.a === 1.0
        ? "rgb(" + c.r + ", " + c.g + ", " + c.b + ")"
        : "rgba(" + c.r + ", " + c.g + ", " + c.b + ", " + c.a + ")";
};
exports.toCSS = toCSS;
/**
 * @category instances
 * @since 0.1.5
 */
exports.Eq = Equals.struct({
    r: number.Eq,
    g: number.Eq,
    b: number.Eq,
    a: number.Eq
});
/**
 * @category instances
 * @since 0.1.5
 */
exports.Show = Sh.struct({
    r: number.Show,
    g: number.Show,
    b: number.Show,
    a: number.Show
});
