// ts-colors/lib/X11.js
"use strict";
/**
 * X11 colors as specified by the CSS3 specification (except black and white).
 * See: https://www.w3.org/TR/css3-color/#svg-color
 *
 * @since 0.1.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ghostwhite = exports.gainsboro = exports.fuchsia = exports.forestgreen = exports.floralwhite = exports.firebrick = exports.dodgerblue = exports.dimgrey = exports.dimgray = exports.deepskyblue = exports.deeppink = exports.darkviolet = exports.darkturquoise = exports.darkslategrey = exports.darkslategray = exports.darkslateblue = exports.darkseagreen = exports.darksalmon = exports.darkred = exports.darkorchid = exports.darkorange = exports.darkolivegreen = exports.darkmagenta = exports.darkkhaki = exports.darkgrey = exports.darkgreen = exports.darkgray = exports.darkgoldenrod = exports.darkcyan = exports.darkblue = exports.cyan = exports.crimson = exports.cornsilk = exports.cornflowerblue = exports.coral = exports.chocolate = exports.chartreuse = exports.cadetblue = exports.burlywood = exports.brown = exports.blueviolet = exports.blue = exports.blanchedalmond = exports.bisque = exports.beige = exports.azure = exports.aquamarine = exports.aqua = exports.antiquewhite = exports.aliceblue = void 0;
exports.navajowhite = exports.moccasin = exports.mistyrose = exports.mintcream = exports.midnightblue = exports.mediumvioletred = exports.mediumturquoise = exports.mediumspringgreen = exports.mediumslateblue = exports.mediumseagreen = exports.mediumpurple = exports.mediumorchid = exports.mediumblue = exports.mediumaquamarine = exports.maroon = exports.magenta = exports.linen = exports.limegreen = exports.lime = exports.lightyellow = exports.lightsteelblue = exports.lightslategrey = exports.lightslategray = exports.lightskyblue = exports.lightseagreen = exports.lightsalmon = exports.lightpink = exports.lightgrey = exports.lightgreen = exports.lightgray = exports.lightgoldenrodyellow = exports.lightcyan = exports.lightcoral = exports.lightblue = exports.lemonchiffon = exports.lawngreen = exports.lavenderblush = exports.lavender = exports.khaki = exports.ivory = exports.indigo = exports.indianred = exports.hotpink = exports.honeydew = exports.grey = exports.greenyellow = exports.green = exports.gray = exports.goldenrod = exports.gold = void 0;
exports.yellowgreen = exports.yellow = exports.whitesmoke = exports.wheat = exports.violet = exports.turquoise = exports.tomato = exports.thistle = exports.teal = exports.tan = exports.steelblue = exports.springgreen = exports.snow = exports.slategrey = exports.slategray = exports.slateblue = exports.skyblue = exports.silver = exports.sienna = exports.seashell = exports.seagreen = exports.sandybrown = exports.salmon = exports.saddlebrown = exports.royalblue = exports.rosybrown = exports.red = exports.purple = exports.powderblue = exports.plum = exports.pink = exports.peru = exports.peachpuff = exports.papayawhip = exports.palevioletred = exports.paleturquoise = exports.palegreen = exports.palegoldenrod = exports.orchid = exports.orangered = exports.orange = exports.olivedrab = exports.olive = exports.oldlace = exports.navy = void 0;
var Color_1 = require("./Color");
/**
 * @since 0.1.0
 */
exports.aliceblue = Color_1.rgb(240, 248, 255);
/**
 * @since 0.1.0
 */
exports.antiquewhite = Color_1.rgb(250, 235, 215);
/**
 * @since 0.1.0
 */
exports.aqua = Color_1.rgb(0, 255, 255);
/**
 * @since 0.1.0
 */
exports.aquamarine = Color_1.rgb(127, 255, 212);
/**
 * @since 0.1.0
 */
exports.azure = Color_1.rgb(240, 255, 255);
/**
 * @since 0.1.0
 */
exports.beige = Color_1.rgb(245, 245, 220);
/**
 * @since 0.1.0
 */
exports.bisque = Color_1.rgb(255, 228, 196);
/**
 * @since 0.1.0
 */
exports.blanchedalmond = Color_1.rgb(255, 235, 205);
/**
 * @since 0.1.0
 */
exports.blue = Color_1.rgb(0, 0, 255);
/**
 * @since 0.1.0
 */
exports.blueviolet = Color_1.rgb(138, 43, 226);
/**
 * @since 0.1.0
 */
exports.brown = Color_1.rgb(165, 42, 42);
/**
 * @since 0.1.0
 */
exports.burlywood = Color_1.rgb(222, 184, 135);
/**
 * @since 0.1.0
 */
exports.cadetblue = Color_1.rgb(95, 158, 160);
/**
 * @since 0.1.0
 */
exports.chartreuse = Color_1.rgb(127, 255, 0);
/**
 * @since 0.1.0
 */
exports.chocolate = Color_1.rgb(210, 105, 30);
/**
 * @since 0.1.0
 */
exports.coral = Color_1.rgb(255, 127, 80);
/**
 * @since 0.1.0
 */
exports.cornflowerblue = Color_1.rgb(100, 149, 237);
/**
 * @since 0.1.0
 */
exports.cornsilk = Color_1.rgb(255, 248, 220);
/**
 * @since 0.1.0
 */
exports.crimson = Color_1.rgb(220, 20, 60);
/**
 * @since 0.1.0
 */
exports.cyan = Color_1.rgb(0, 255, 255);
/**
 * @since 0.1.0
 */
exports.darkblue = Color_1.rgb(0, 0, 139);
/**
 * @since 0.1.0
 */
exports.darkcyan = Color_1.rgb(0, 139, 139);
/**
 * @since 0.1.0
 */
exports.darkgoldenrod = Color_1.rgb(184, 134, 11);
/**
 * @since 0.1.0
 */
exports.darkgray = Color_1.rgb(169, 169, 169);
/**
 * @since 0.1.0
 */
exports.darkgreen = Color_1.rgb(0, 100, 0);
/**
 * @since 0.1.0
 */
exports.darkgrey = Color_1.rgb(169, 169, 169);
/**
 * @since 0.1.0
 */
exports.darkkhaki = Color_1.rgb(189, 183, 107);
/**
 * @since 0.1.0
 */
exports.darkmagenta = Color_1.rgb(139, 0, 139);
/**
 * @since 0.1.0
 */
exports.darkolivegreen = Color_1.rgb(85, 107, 47);
/**
 * @since 0.1.0
 */
exports.darkorange = Color_1.rgb(255, 140, 0);
/**
 * @since 0.1.0
 */
exports.darkorchid = Color_1.rgb(153, 50, 204);
/**
 * @since 0.1.0
 */
exports.darkred = Color_1.rgb(139, 0, 0);
/**
 * @since 0.1.0
 */
exports.darksalmon = Color_1.rgb(233, 150, 122);
/**
 * @since 0.1.0
 */
exports.darkseagreen = Color_1.rgb(143, 188, 143);
/**
 * @since 0.1.0
 */
exports.darkslateblue = Color_1.rgb(72, 61, 139);
/**
 * @since 0.1.0
 */
exports.darkslategray = Color_1.rgb(47, 79, 79);
/**
 * @since 0.1.0
 */
exports.darkslategrey = Color_1.rgb(47, 79, 79);
/**
 * @since 0.1.0
 */
exports.darkturquoise = Color_1.rgb(0, 206, 209);
/**
 * @since 0.1.0
 */
exports.darkviolet = Color_1.rgb(148, 0, 211);
/**
 * @since 0.1.0
 */
exports.deeppink = Color_1.rgb(255, 20, 147);
/**
 * @since 0.1.0
 */
exports.deepskyblue = Color_1.rgb(0, 191, 255);
/**
 * @since 0.1.0
 */
exports.dimgray = Color_1.rgb(105, 105, 105);
/**
 * @since 0.1.0
 */
exports.dimgrey = Color_1.rgb(105, 105, 105);
/**
 * @since 0.1.0
 */
exports.dodgerblue = Color_1.rgb(30, 144, 255);
/**
 * @since 0.1.0
 */
exports.firebrick = Color_1.rgb(178, 34, 34);
/**
 * @since 0.1.0
 */
exports.floralwhite = Color_1.rgb(255, 250, 240);
/**
 * @since 0.1.0
 */
exports.forestgreen = Color_1.rgb(34, 139, 34);
/**
 * @since 0.1.0
 */
exports.fuchsia = Color_1.rgb(255, 0, 255);
/**
 * @since 0.1.0
 */
exports.gainsboro = Color_1.rgb(220, 220, 220);
/**
 * @since 0.1.0
 */
exports.ghostwhite = Color_1.rgb(248, 248, 255);
/**
 * @since 0.1.0
 */
exports.gold = Color_1.rgb(255, 215, 0);
/**
 * @since 0.1.0
 */
exports.goldenrod = Color_1.rgb(218, 165, 32);
/**
 * @since 0.1.0
 */
exports.gray = Color_1.rgb(128, 128, 128);
/**
 * @since 0.1.0
 */
exports.green = Color_1.rgb(0, 128, 0);
/**
 * @since 0.1.0
 */
exports.greenyellow = Color_1.rgb(173, 255, 47);
/**
 * @since 0.1.0
 */
exports.grey = Color_1.rgb(128, 128, 128);
/**
 * @since 0.1.0
 */
exports.honeydew = Color_1.rgb(240, 255, 240);
/**
 * @since 0.1.0
 */
exports.hotpink = Color_1.rgb(255, 105, 180);
/**
 * @since 0.1.0
 */
exports.indianred = Color_1.rgb(205, 92, 92);
/**
 * @since 0.1.0
 */
exports.indigo = Color_1.rgb(75, 0, 130);
/**
 * @since 0.1.0
 */
exports.ivory = Color_1.rgb(255, 255, 240);
/**
 * @since 0.1.0
 */
exports.khaki = Color_1.rgb(240, 230, 140);
/**
 * @since 0.1.0
 */
exports.lavender = Color_1.rgb(230, 230, 250);
/**
 * @since 0.1.0
 */
exports.lavenderblush = Color_1.rgb(255, 240, 245);
/**
 * @since 0.1.0
 */
exports.lawngreen = Color_1.rgb(124, 252, 0);
/**
 * @since 0.1.0
 */
exports.lemonchiffon = Color_1.rgb(255, 250, 205);
/**
 * @since 0.1.0
 */
exports.lightblue = Color_1.rgb(173, 216, 230);
/**
 * @since 0.1.0
 */
exports.lightcoral = Color_1.rgb(240, 128, 128);
/**
 * @since 0.1.0
 */
exports.lightcyan = Color_1.rgb(224, 255, 255);
/**
 * @since 0.1.0
 */
exports.lightgoldenrodyellow = Color_1.rgb(250, 250, 210);
/**
 * @since 0.1.0
 */
exports.lightgray = Color_1.rgb(211, 211, 211);
/**
 * @since 0.1.0
 */
exports.lightgreen = Color_1.rgb(144, 238, 144);
/**
 * @since 0.1.0
 */
exports.lightgrey = Color_1.rgb(211, 211, 211);
/**
 * @since 0.1.0
 */
exports.lightpink = Color_1.rgb(255, 182, 193);
/**
 * @since 0.1.0
 */
exports.lightsalmon = Color_1.rgb(255, 160, 122);
/**
 * @since 0.1.0
 */
exports.lightseagreen = Color_1.rgb(32, 178, 170);
/**
 * @since 0.1.0
 */
exports.lightskyblue = Color_1.rgb(135, 206, 250);
/**
 * @since 0.1.0
 */
exports.lightslategray = Color_1.rgb(119, 136, 153);
/**
 * @since 0.1.0
 */
exports.lightslategrey = Color_1.rgb(119, 136, 153);
/**
 * @since 0.1.0
 */
exports.lightsteelblue = Color_1.rgb(176, 196, 222);
/**
 * @since 0.1.0
 */
exports.lightyellow = Color_1.rgb(255, 255, 224);
/**
 * @since 0.1.0
 */
exports.lime = Color_1.rgb(0, 255, 0);
/**
 * @since 0.1.0
 */
exports.limegreen = Color_1.rgb(50, 205, 50);
/**
 * @since 0.1.0
 */
exports.linen = Color_1.rgb(250, 240, 230);
/**
 * @since 0.1.0
 */
exports.magenta = Color_1.rgb(255, 0, 255);
/**
 * @since 0.1.0
 */
exports.maroon = Color_1.rgb(128, 0, 0);
/**
 * @since 0.1.0
 */
exports.mediumaquamarine = Color_1.rgb(102, 205, 170);
/**
 * @since 0.1.0
 */
exports.mediumblue = Color_1.rgb(0, 0, 205);
/**
 * @since 0.1.0
 */
exports.mediumorchid = Color_1.rgb(186, 85, 211);
/**
 * @since 0.1.0
 */
exports.mediumpurple = Color_1.rgb(147, 112, 219);
/**
 * @since 0.1.0
 */
exports.mediumseagreen = Color_1.rgb(60, 179, 113);
/**
 * @since 0.1.0
 */
exports.mediumslateblue = Color_1.rgb(123, 104, 238);
/**
 * @since 0.1.0
 */
exports.mediumspringgreen = Color_1.rgb(0, 250, 154);
/**
 * @since 0.1.0
 */
exports.mediumturquoise = Color_1.rgb(72, 209, 204);
/**
 * @since 0.1.0
 */
exports.mediumvioletred = Color_1.rgb(199, 21, 133);
/**
 * @since 0.1.0
 */
exports.midnightblue = Color_1.rgb(25, 25, 112);
/**
 * @since 0.1.0
 */
exports.mintcream = Color_1.rgb(245, 255, 250);
/**
 * @since 0.1.0
 */
exports.mistyrose = Color_1.rgb(255, 228, 225);
/**
 * @since 0.1.0
 */
exports.moccasin = Color_1.rgb(255, 228, 181);
/**
 * @since 0.1.0
 */
exports.navajowhite = Color_1.rgb(255, 222, 173);
/**
 * @since 0.1.0
 */
exports.navy = Color_1.rgb(0, 0, 128);
/**
 * @since 0.1.0
 */
exports.oldlace = Color_1.rgb(253, 245, 230);
/**
 * @since 0.1.0
 */
exports.olive = Color_1.rgb(128, 128, 0);
/**
 * @since 0.1.0
 */
exports.olivedrab = Color_1.rgb(107, 142, 35);
/**
 * @since 0.1.0
 */
exports.orange = Color_1.rgb(255, 165, 0);
/**
 * @since 0.1.0
 */
exports.orangered = Color_1.rgb(255, 69, 0);
/**
 * @since 0.1.0
 */
exports.orchid = Color_1.rgb(218, 112, 214);
/**
 * @since 0.1.0
 */
exports.palegoldenrod = Color_1.rgb(238, 232, 170);
/**
 * @since 0.1.0
 */
exports.palegreen = Color_1.rgb(152, 251, 152);
/**
 * @since 0.1.0
 */
exports.paleturquoise = Color_1.rgb(175, 238, 238);
/**
 * @since 0.1.0
 */
exports.palevioletred = Color_1.rgb(219, 112, 147);
/**
 * @since 0.1.0
 */
exports.papayawhip = Color_1.rgb(255, 239, 213);
/**
 * @since 0.1.0
 */
exports.peachpuff = Color_1.rgb(255, 218, 185);
/**
 * @since 0.1.0
 */
exports.peru = Color_1.rgb(205, 133, 63);
/**
 * @since 0.1.0
 */
exports.pink = Color_1.rgb(255, 192, 203);
/**
 * @since 0.1.0
 */
exports.plum = Color_1.rgb(221, 160, 221);
/**
 * @since 0.1.0
 */
exports.powderblue = Color_1.rgb(176, 224, 230);
/**
 * @since 0.1.0
 */
exports.purple = Color_1.rgb(128, 0, 128);
/**
 * @since 0.1.0
 */
exports.red = Color_1.rgb(255, 0, 0);
/**
 * @since 0.1.0
 */
exports.rosybrown = Color_1.rgb(188, 143, 143);
/**
 * @since 0.1.0
 */
exports.royalblue = Color_1.rgb(65, 105, 225);
/**
 * @since 0.1.0
 */
exports.saddlebrown = Color_1.rgb(139, 69, 19);
/**
 * @since 0.1.0
 */
exports.salmon = Color_1.rgb(250, 128, 114);
/**
 * @since 0.1.0
 */
exports.sandybrown = Color_1.rgb(244, 164, 96);
/**
 * @since 0.1.0
 */
exports.seagreen = Color_1.rgb(46, 139, 87);
/**
 * @since 0.1.0
 */
exports.seashell = Color_1.rgb(255, 245, 238);
/**
 * @since 0.1.0
 */
exports.sienna = Color_1.rgb(160, 82, 45);
/**
 * @since 0.1.0
 */
exports.silver = Color_1.rgb(192, 192, 192);
/**
 * @since 0.1.0
 */
exports.skyblue = Color_1.rgb(135, 206, 235);
/**
 * @since 0.1.0
 */
exports.slateblue = Color_1.rgb(106, 90, 205);
/**
 * @since 0.1.0
 */
exports.slategray = Color_1.rgb(112, 128, 144);
/**
 * @since 0.1.0
 */
exports.slategrey = Color_1.rgb(112, 128, 144);
/**
 * @since 0.1.0
 */
exports.snow = Color_1.rgb(255, 250, 250);
/**
 * @since 0.1.0
 */
exports.springgreen = Color_1.rgb(0, 255, 127);
/**
 * @since 0.1.0
 */
exports.steelblue = Color_1.rgb(70, 130, 180);
/**
 * @since 0.1.0
 */
exports.tan = Color_1.rgb(210, 180, 140);
/**
 * @since 0.1.0
 */
exports.teal = Color_1.rgb(0, 128, 128);
/**
 * @since 0.1.0
 */
exports.thistle = Color_1.rgb(216, 191, 216);
/**
 * @since 0.1.0
 */
exports.tomato = Color_1.rgb(255, 99, 71);
/**
 * @since 0.1.0
 */
exports.turquoise = Color_1.rgb(64, 224, 208);
/**
 * @since 0.1.0
 */
exports.violet = Color_1.rgb(238, 130, 238);
/**
 * @since 0.1.0
 */
exports.wheat = Color_1.rgb(245, 222, 179);
/**
 * @since 0.1.0
 */
exports.whitesmoke = Color_1.rgb(245, 245, 245);
/**
 * @since 0.1.0
 */
exports.yellow = Color_1.rgb(255, 255, 0);
/**
 * @since 0.1.0
 */
exports.yellowgreen = Color_1.rgb(154, 205, 50);
